source/dub/internal/utils.d

```diff
@@ -342,6 +342,63 @@ string getDUBVersion()
 	return verstr;
 }
 
+
+/**
+	Get current executable's path if running as DUB executable,
+	or find a DUB executable if DUB is used as a library.
+	For the latter, the following locations are checked in order:
+	$(UL
+		$(LI current working directory)
+		$(LI same directory as `compilerBinary` (if supplied))
+		$(LI all components of the `$PATH` variable)
+	)
+	Params:
+		compilerBinary = optional path to a D compiler executable, used to locate DUB executable
+	Returns:
+		The path to a valid DUB executable
+	Throws:
+		an Exception if no valid DUB executable is found
+*/
+public string getDUBExePath(in string compilerBinary=null)
+{
+	version(DubApplication) {
+		import std.file : thisExePath;
+		return thisExePath();
+	}
+	else {
+		// this must be dub as a library
+		import std.algorithm : filter, map, splitter;
+		import std.file : exists, getcwd;
+		import std.path : buildPath, dirName;
+		import std.range : chain, only;
+		import std.process : environment;
+
+		version(Windows) {
+			enum exeName = "dub.exe";
+			enum pathSep = ';';
+		}
+		else {
+			enum exeName = "dub";
+			enum pathSep = ':';
+		}
+
+		auto dubLocs = only(
+			getcwd().buildPath(exeName),
+			compilerBinary ? compilerBinary.dirName.buildPath(exeName) : null,
```

Comment:
you should be doing `compilerBinary.length ? ...` here instead so if an empty string is passed it won't try to find the exe there.

```suggestion
			compilerBinary.length ? compilerBinary.dirName.buildPath(exeName) : null,
```